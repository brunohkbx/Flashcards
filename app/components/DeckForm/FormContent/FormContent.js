import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { ScrollView } from 'react-native';
import { Headline } from 'react-native-paper';
import { Field } from 'formik';
import FormInput from '../FormInput/FormInput';
import Flashcard from '../../Flashcard';
import { Fadable } from '../../Animations';

class FormContent extends Component {
  state = {
    flashcardToRemove: null
  }

  deleteDeck = index => {
    const { onFlashcardDeleted } = this.props;

    this.setState({ flashcardToRemove: null }, () => onFlashcardDeleted(index));
  };

  render() {
    const { values, handleChange } = this.props;
    const { flashcardToRemove } = this.state;

    return (
      <ScrollView>
        <Field
          name="title"
          label="Title"
          component={FormInput}
          onChangeText={handleChange('title')}
          autoFocus
        />
        <Headline>Cards</Headline>
        {
          values.questions.map((question, index) => (
            <Fadable
              key={question.id}
              fade={flashcardToRemove === index}
              onAnimationEnd={() => this.deleteDeck(index)}
            >
              <Flashcard
                onFlashcardDeleted={
                  () => this.setState({ flashcardToRemove: index })
                }
              >
                <Field
                  name={`questions.${index}.question`}
                  label="Question"
                  component={FormInput}
                  onChangeText={handleChange(`questions.${index}.question`)}
                  autoFocus
                  multiline
                  mode="outlined"
                />
                <Field
                  name={`questions.${index}.answer`}
                  label="Answer"
                  component={FormInput}
                  onChangeText={handleChange(`questions.${index}.answer`)}
                  multiline
                  mode="outlined"
                />
              </Flashcard>
            </Fadable>
          ))
        }
      </ScrollView>
    );
  }
};

FormContent.propTypes = {
  handleChange: PropTypes.func.isRequired,
  values: PropTypes.object.isRequired,
  onFlashcardDeleted: PropTypes.func.isRequired
};

export default FormContent;
