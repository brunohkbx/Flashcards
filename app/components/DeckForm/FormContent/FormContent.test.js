import React from 'react';
import { shallow } from 'enzyme';
import FormContent from './FormContent';

describe('FormContent', () => {
  const setup = propOverrides => {
    const defaultProps = Object.assign({
      handleChange: jest.fn(),
      values: {
        title: 'React',
        questions: [{
          question: 'What is React?',
          answer: 'A library for managing user interfaces'
        }],
        id: '1'
      },
      handleContentSizeChange: jest.fn()
    }, propOverrides);

    const wrapper = shallow(<FormContent {...defaultProps} />);

    const contentSizeChange = () => {
      wrapper.find('ScrollView').simulate('contentSizeChange');
    };

    return {
      wrapper,
      contentSizeChange
    };
  };

  it('renders properly', () => {
    const { wrapper } =  setup();

    expect(wrapper).toMatchSnapshot();
  });

  test('ScrollView onContentSizeChange calls handleContentSizeChange', () => {
    const mockHandleContentSizeChange = jest.fn();
    const { contentSizeChange } = setup(
      { handleContentSizeChange: mockHandleContentSizeChange }
    );

    contentSizeChange();

    expect(mockHandleContentSizeChange).toHaveBeenCalled();
  });
});
