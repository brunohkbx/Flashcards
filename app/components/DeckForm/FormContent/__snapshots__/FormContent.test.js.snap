// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormContent renders properly 1`] = `
<ScrollView>
  <C
    autoFocus={true}
    component={[Function]}
    label="Title"
    name="title"
  />
  <Headline>
    Cards
  </Headline>
  <Fadable
    fade={false}
    key="1"
    onAnimationEnd={[Function]}
  >
    <Flashcard
      onFlashcardDeleted={[Function]}
    >
      <C
        autoFocus={true}
        component={[Function]}
        label="Question"
        mode="outlined"
        multiline={true}
        name="questions.0.question"
      />
      <C
        component={[Function]}
        label="Answer"
        mode="outlined"
        multiline={true}
        name="questions.0.answer"
      />
    </Flashcard>
  </Fadable>
</ScrollView>
`;
